class/practice2.2/main.cpp

```diff
@@ -0,0 +1,204 @@
+    }
+
+    bool isValid() const {
+        return (color.r >= 0 && color.r <= 255) &&
+            (color.g >= 0 && color.g <= 255) &&
+            (color.b >= 0 && color.b <= 255);
+    }
+
+    bool compareColor(const Pixel& compare) const {
+        return (compare.color.b==color.b) && (compare.color.g==color.g) && (compare.color.r==color.r);
+    }
+
+    void setColor(const Pixel& col){
+         color = col.getColor();
+    }
+
+private:
+
+    int convertToInt(const std::string& str) {
+        try {
+            return static_cast<int>(std::stoi(str));
+        } catch (const std::invalid_argument& ia) {
+            return -1;
+        } catch (const std::out_of_range& oor) {
+            return -1;
+        }
+    }
+    Color color;
+};
+
+std::string listToStr(const std::vector<Pixel>& list) {
+    std::ostringstream oss;
+
+    for (const Pixel& pix : list) {
+        if (!oss.str().empty()) {
+            oss << " ";
+        }
+        oss << pix.serToString();
+    }
+    return oss.str();
+}
+
+class Image {
+private:
+    std::vector<std::vector<Pixel>> lines;
+    std::string pixelMatrixFile;
+
+public:
+
+    Image(const std::string& filePath) : pixelMatrixFile(filePath), lines(MatrixSize, std::vector<Pixel>(MatrixSize)) {}
+
+    int loadMatrix() {
+        std::string line;
+        std::ifstream pixelMatrix(pixelMatrixFile);
+        int indexCounter=0;
+        while (std::getline(pixelMatrix, line)) {
+            std::vector<std::string>  splittedLine = splitString(line, ' ');
+            std::vector<Pixel> pixels(MatrixSize);
+             if (splittedLine.size() != MatrixSize)
+            {
+                return 1;
+            }
+            for (int i = 0; i < MatrixSize; i++)
+            {
+                Pixel pix(splittedLine[i]);
+                if (!pix.isValid()) {
+                    return 2;
+                }
+                pixels[i]=pix;
+            }
+
+            if (indexCounter==MatrixSize)
+            {
+                return 1;
+            }
+            
+            lines[indexCounter]=pixels;
+            indexCounter++;
+        }
+        pixelMatrix.close();
+        return 0;
+    }
+ 
+    void changeMatrix(const Pixel& favouriteColor) {
+        for (int i = 0; i < MatrixSize; i++)
+        {
+            for (int j = 0; j < MatrixSize; j++)
+            {
+                if (lines[i][j].compareColor(favouriteColor)) {
+                    if (i > 0)
+                    {
+                        lines[i - 1][j].setColor(favouriteColor);
+                    }
+                    if (j > 0)
+                    {
+                        lines[i][j - 1].setColor(favouriteColor);
+                    }
+                }
+            }
+        }
+    }
+
+    int saveFile() {
+      std::ofstream fileEdit("Output" + pixelMatrixFile, std::ios::trunc);
+      for (int i = 0; i < MatrixSize; i++)
+      {
+          std::string convertedList = listToStr(lines[i]);
```

Comment:
we should not have empty lines as we validated that image is correct. And therefore it can be written like
```
void saveFile() {
    std::ofstream fileEdit("Output" + pixelMatrixFile, std::ios::trunc);
    for (const auto& line : lines) {
        std::println(fileEdit, listToStr(line));
    }
}
```

there is no need to have return value of int if you don't have any other option except return 0